dallinger/experiment.py

```diff
@@ -511,16 +511,18 @@ def networks(self, role="all", full="all"):
 
         if full == "all":
             if role == "all":
-                return Network.query.all()
```

Comment:
Trying to understand the rationale for `Experiment.session`. Is the point just performance optimisation (we ask `scoped_session` to fetch us a `session` and we store that in `Experiment` to avoid repeatedly querying `scoped_session`?) Is this optimisation really necessary? It seems confusing to me to have this alternative route of accessing session.

Reply:
I feel like if we're not careful this is another route to tripping ourselves up? Isn't it safer to refer to the original scoped session every time?

Reply:
@pmcharrison In the end I guess we didn't resolve this beyond establishing that there's no _harm_ in using `self.session`, provided it's always a reference to the scoped_session registry, and not an actual session instance.

Given everything else we've learned, the only downside I see to doing away with the session instance reference is that if we ever wanted to move _away_ from using the global, we'd have more work to do. Well, also: if tests want to fake the database for some reason, they have to monkeypatch the module instead of passing a fake session, but I don't think we're doing any of that currently.

Do you think doing away with `self.session` would feel simpler?

@alecpm thoughts on this?

Reply:
@jessesnyder I like having `self.session`, using what looks like a global session in every method seems weird, but it doesn't really matter. The difference is just stylistic as far as I can tell. I think currently `exp.session` is always available when you're running an experiment. I think the only time `session` isn't passed to the constructor is when the command line stuff is trying to introspect the experiment, load data, etc., in which case it doesn't want to run `configure`/`setup` and create networks, etc. I think implicitly relying on `session` for that is a bit weird, but making `session` required will probably break something if we don't have a more explicit mechanism.

Reply:
I made a change so that session is always required, missing:
1. the scenarios where we need to instantiate experiments with no session, and 
2. use the presence of session as a flag to indicate whether to run setup

I think we should probably revert my change here, though I don't like all the code that assumes `self.session` is set, when it might be None. It seems like the calls to experment that don't use a need a session should be classmethods or something.

Reply:
In the long term `classmethod` does seem like the right approach for things like `Experiment.run`, `Experiment.collect`, and `Experiment.jupyter_replay`.

Reply:
I've reverted the change so that session is again not required. We can continue to use this sub-optimal flag for whether to run setup until we decide we want something better.